infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/Names.java

```diff
@@ -98,4 +162,83 @@ public static String uniqueResourceName(String originalResourceName, String work
   public static String generateName(String prefix) {
     return NameGenerator.generate(prefix, GENERATED_PART_SIZE);
   }
+
+  private static String findMachineName(Map<String, String> annotations, String containerName) {
+    if (annotations == null) {
+      return null;
+    }
+
+    for (Map.Entry<String, String> e : annotations.entrySet()) {
+      String key = e.getKey();
+
+      if (!key.startsWith(CONTAINER_META_PREFIX)) {
+        continue;
+      }
+
+      if (!key.endsWith(CONTAINER_META_NAME_SUFFIX)) {
+        continue;
+      }
+
+      if (!containerName.equals(e.getValue())) {
+        continue;
+      }
+
+      String index =
+          key.substring(
+              CONTAINER_META_PREFIX.length(), key.length() - CONTAINER_META_NAME_SUFFIX.length());
+
+      String machineNameKey = CONTAINER_META_PREFIX + index + CONTAINER_META_MACHINE_SUFFIX;
+
+      return annotations.get(machineNameKey);
+    }
+
+    return null;
+  }
+
+  private static int maxContainerMetaIndex(Map<String, String> annotations) {
+    int max = 0;
+
+    if (annotations == null) {
+      return max;
+    }
+
+    for (String key : annotations.keySet()) {
+      if (!(key.startsWith(CONTAINER_META_PREFIX) && key.endsWith(CONTAINER_META_NAME_SUFFIX))) {
```

Comment:
What do you think about using Pattern instead of startsWith+endsWith+substring?
Like the following
https://gist.github.com/sleshchenko/f2190306fdd7c9dc2cbb071bc22d64e0